lib/flight_job/matcher.rb

```diff
@@ -0,0 +1,65 @@
+#==============================================================================
+# Copyright (C) 2022-present Alces Flight Ltd.
+#
+# This file is part of Flight Job.
+#
+# This program and the accompanying materials are made available under
+# the terms of the Eclipse Public License 2.0 which is available at
+# <https://www.eclipse.org/legal/epl-2.0>, or alternative license
+# terms made available by Alces Flight Ltd - please direct inquiries
+# about licensing to licensing@alces-flight.com.
+#
+# FlightHowto is distributed in the hope that it will be useful, but
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, EITHER EXPRESS OR
+# IMPLIED INCLUDING, WITHOUT LIMITATION, ANY WARRANTIES OR CONDITIONS
+# OF TITLE, NON-INFRINGEMENT, MERCHANTABILITY OR FITNESS FOR A
+# PARTICULAR PURPOSE. See the Eclipse Public License 2.0 for more
+# details.
+#
+# You should have received a copy of the Eclipse Public License 2.0
+# along with FlightHowto. If not, see:
+#
+#  https://opensource.org/licenses/EPL-2.0
+#
+# For more information on FlightHowto, please visit:
+# https://github.com/openflighthpc/flight-howto
+#==============================================================================
+module FlightJob
+  class Matcher
+    attr_reader :filters, :attrs
```

Comment:
You've ran into one of my pet peeves with Ruby here: overuse of `attr_reader` (and more generally `attr_accessor`).  I'll admit that it's very common to create reader (and accessor) methods for instance variables, but I don't like it.  I don't like it because code is easier to refactor the smaller its public interface is.

For instance, when the interface for `Matcher` is simply `Matcher::new` and `Matcher#matches?` it is (1) very clear how that class can be used; and (2) as long as those two methods continue to work unchanged, we can make as many changes to its internals as we wish knowing that none of the `Matcher` clients will need to be updated.

If internals such as `@filters` and `@attrs` are made public through `attr_reader` then even something as simple as renaming `attrs` has the potential to break clients.

Please either use instance variables directly or make the readers private by moving them below the `private` keyword.